lib/bundle/installer.rb

```diff
@@ -5,47 +5,121 @@ module Installer
     module_function
 
     def install(entries)
-      success = 0
-      failure = 0
-
-      entries.each do |entry|
-        arg = [entry.name]
-        verb = "Installing"
-        cls = case entry.type
-        when :brew
-          arg << entry.options
-          Bundle::BrewInstaller
-        when :cask
-          arg << entry.options
-          Bundle::CaskInstaller
-        when :mac_app_store
-          arg << entry.options[:id]
-          Bundle::MacAppStoreInstaller
-        when :tap
-          verb = "Tapping"
-          arg << entry.options
-          Bundle::TapInstaller
-        end
-        case cls.install(*arg)
-        when :success
-          puts Formatter.success("#{verb} #{entry.name}")
-          success += 1
-        when :skipped
-          puts "Using #{entry.name}"
-          success += 1
+      end_state = install_entries(entries)
+
+      failures = end_state[:failed]
+      report_success_and_failure(end_state)
+
+      errored_entries = end_state[:errored]
+      report_errors(errored_entries) if errored_entries.any?
+
+      failures.zero? && errored_entries.empty?
```

Comment:
Why make an (internal) API where you're checking for `zero?` for one and `empty?` for the other?

Reply:
Until now, the successes, skipped, and failures have only been tracked as a count instead of the formula name and maybe options. Errors need that metadata in order to be displayed later.

I'd welcome keeping more information about successes, skipped, and failure even if that information isn't used.

Maybe there's insufficient dissimilarity between failures and errors that I'm missing in how I've approached the problem?